### PR TITLE
Using streaming API for deserialization

### LINKED ISSUE #38725: Using stream-style-serialization: true for Assistants

### PR DESCRIPTION
**Pull request by jpalvarezl**
I am basing the migration on this [guide](https://github.com/Azure/azure-sdk-for-java/wiki/Azure-Json-Migration)

Closes #38725

### Current state

All issues with the emitter have been solved. We are waiting for this [PR](https://github.com/Azure/azure-rest-api-specs/pull/28665) to go into `main` to update the commit hash and run code gen again.

### Keeping the issues here documentation sake
- Having issues with dates as `long`
  - Using `ThreadRun.java` as an example. Here is [TSP](https://github.com/Azure/azure-rest-api-specs/blob/ecc0170a2005f5f38231ae4dbba40594d3c00a04/specification/ai/OpenAI.Assistants/runs/models.tsp#L70) definition. It appears that making the date nullable changes the way this is parsed.
```java 
// .... 
} else if ("created_at".equals(fieldName)) {
  createdAt = OffsetDateTime.ofInstant(Instant.ofEpochSecond(reader.getLong()), ZoneOffset.UTC); // <-- Works!
} else if ("expires_at".equals(fieldName)) {
  expiresAt = reader.getNullable(nonNullReader -> OffsetDateTime.parse(nonNullReader.getString())); //  <-- Throws!
```
- Issues with generic type parameter of `PageableList<T>`
  - Using Accessor pattern to sidestep generic type parameter deserialization, based on this [guide](https://github.com/Azure/azure-sdk-for-java/wiki/Access-helpers)
- Polymorphic types:
  - `RetrievalToolCall`
  - `CodeInterpreterImageOutput`
  - `CodeInterpreterImageReference`
  - `CodeInterpreterLogOutput`
  - `CodeInterpreterToolCall`
  - `CodeInterpreterToolCallDetails`
  - `CodeInterpreterToolCallOuput`
  - `ToolCall`
  - And more ... just do a project-wide search for `com.fasterxml.jackson`

### CODE REVIEW

File: sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/AssistantsAsyncClient.java [hunk 1]
```diff
@@ -3043,8 +3109,53 @@ public Mono<OpenAIFile> uploadFile(FileDetails file, FilePurpose purpose) {
...
+     * @param requestOptions The options to configure the HTTP request before HTTP client sends it.
+     * @throws HttpResponseException thrown if the request is rejected by server.
+     * @throws ClientAuthenticationException thrown if the request is rejected by server on status code 401.
+     * @throws ResourceNotFoundException thrown if the request is rejected by server on status code 404.
+     * @throws ResourceModifiedException thrown if the request is rejected by server on status code 409.
+     * @return represent a byte array along with {@link Response} on successful completion of {@link Mono}.
+     */
+    @Generated
+    @ServiceMethod(returns = ReturnType.SINGLE)
+    public Mono<Response<BinaryData>> getFileContentWithResponse(String fileId, RequestOptions requestOptions) {
```
Comment (mssfang) [inline sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/AssistantsAsyncClient.java:3137]: We should add a new feature entry in the CHANGELOG

**Review by mssfang — APPROVED** [review]

File: sdk/openai/azure-ai-openai-assistants/src/main/java/module-info.java [hunk 2]
```diff
@@ -4,9 +4,11 @@
 
 module com.azure.ai.openai.assistants {
     requires transitive com.azure.core;
+    requires transitive com.azure.json;
+    requires jsr305;
```
[the remark this answers is not in the document]
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai-assistants/src/main/java/module-info.java:8]: Yeah, that somehow appeared on its own after the last updates to the emitter🤔

[the remark this answers is not in the document]
Comment (weidongxu-microsoft) [reply sdk/openai/azure-ai-openai-assistants/src/main/java/module-info.java:8]: I couldn't find it in autorest.java source
image

Maybe try delete this line and regen?

**Comment by jpalvarezl** [comment]
@mssfang am I missing something or the tests that are failing are from Inference and not from Assistants?

File: sdk/openai/azure-ai-openai-assistants/src/main/java/module-info.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai-assistants/src/main/java/module-info.java]: That did it, after deleting it didn't re-appear after regen 👍

File: sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/AssistantsAsyncClient.java (hunk 1 above)
Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/AssistantsAsyncClient.java:3137]: Oh sorry I didn't realise this was about new stuff. I thought it was just codegen related convenience methods. In that case I agree with you on maybe punting this into a separate issue.

Comment (jpalvarezl) [reply sdk/openai/azure-ai-openai-assistants/src/main/java/com/azure/ai/openai/assistants/AssistantsAsyncClient.java:3137]: Issue created: https://github.com/Azure/azure-sdk-for-java/issues/39675

**Pull request merged**